File: python/cycle_detection.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def create_cycle(self, pos):
        """Creates a cycle in the linked list at the given position"""
        if pos < 0:
            return
        cycle_node = self.head
        last_node = self.head
        count = 0
        while count < pos and cycle_node:
            cycle_node = cycle_node.next
            count += 1
        while last_node.next:
            last_node = last_node.next
        last_node.next = cycle_node

    def detect_cycle(self):
        """Floyd’s Cycle Detection Algorithm (Tortoise and Hare)"""
        slow = self.head
        fast = self.head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False


# Example usage:
if __name__ == "__main__":
    ll = LinkedList()
    ll.append(1)
    ll.append(2)
    ll.append(3)
    ll.append(4)
    ll.append(5)

    print("Cycle detected?" , ll.detect_cycle())  # False

    ll.create_cycle(2)  # Create cycle at node with value 3
    print("Cycle detected?" , ll.detect_cycle())  # True

File: c++/Linked Lists/cycle_detection.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};

bool hasCycle(Node* head) {
    Node *slow = head, *fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast)
            return true;
    }
    return false;
}

int main() {
    Node* head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->next->next->next = new Node(4);

    // Create a cycle for testing: point last node to second node
    head->next->next->next->next = head->next;

    if (hasCycle(head))
        cout << "Cycle detected\n";
    else
        cout << "No cycle\n";

    return 0;
}
